### ISSUE TITLE
Is there a LZ4 streaming compression API comparable to those in zstd and zlib ?

### ISSUE THREAD

**Issue by bharatnc** [lz4/lz4#1009]
I'm looking for an API in LZ4 that's comparable to those offered by zstd or zlib for streaming compression of a file. 

TLDR; I'm looking for an API that will enable me to:
- determine if input stream in available (empty or non-empty)
- keep track of the number bytes written to out buffer
- current position while reading the input data

So that I know when I can stop compression.

Examples:
I'm trying to acheive something similar to the following examples done using zstd and zlib but using lz4:

In zstd I can do something equivalent to:
Full source: https://github.com/facebook/zstd/blob/dev/examples/streaming_compression.c#L64-L77
```
        do {
            /* Compress into the output buffer and write all of the output to
             * the file so we can reuse the buffer next iteration.
             */
            ZSTD_outBuffer output = { buffOut, buffOutSize, 0 };
            size_t const remaining = ZSTD_compressStream2(cctx, &output , &input, mode);
            CHECK_ZSTD(remaining);
            fwrite_orDie(buffOut, output.pos, fout);
            /* If we're on the last chunk we're finished when zstd returns 0,
             * which means its consumed all the input AND finished the frame.
             * Otherwise, we're finished when we've consumed all the input.
             */
            finished = lastChunk ? (remaining == 0) : (input.pos == input.size);
        } while (!finished);
```
Similarly In zlib:
Full source: https://stackoverflow.com/questions/4538586/how-to-compress-a-buffer-with-zlib
```
    z_stream strm;
    strm.zalloc = 0;
    strm.zfree = 0;
    strm.next_in = reinterpret_cast<uint8_t *>(in_data);
    strm.avail_in = in_data_size;
    strm.next_out = temp_buffer;
    strm.avail_out = BUFSIZE;
    
    deflateInit(&strm, Z_BEST_COMPRESSION);
    
    while (strm.avail_in != 0)
    {
    int res = deflate(&strm, Z_NO_FLUSH);
    ...
    }
```
Both if these API's mutate the input context and provide information about avail_in, avail_out etc throughout the duration of the compression, so that I know when to stop compression.

So far I have looked at `LZ4_compress_continue` and `lz4_compress_destsize` and while the later atleast gives me the number of bytes written to destination using `*srcSizePtr`, I'm not able to find an API that is similar to the kind of api I'm looking for or perhaps I'm wrong to compare lz4 compression API's directly to zstd or zlib in which case please excuse me. 

Referred: https://github.com/lz4/lz4/blob/dev/examples/blockStreaming_lineByLine.c

Any pointers will be really helpful. Thanks in advance!

**Comment by t-mat** [comment]
Hi @bharatnc,
Perhaps, you can start with the following example and `LZ4F_compressUpdate()`.

https://github.com/lz4/lz4/blob/c240126068b780d927c126bb6a16cdeb92c35d77/examples/frameCompress.c#L56-L120

I'm not sure you need to use LZ4(HC) instead of LZ4Frame or not.  But please note

(1) LZ4 and LZ4HC provide low level APIs.  (Analogue of DEFLATE/INFLATE raw functions)
(2) LZ4 Frame provides relatively high level API.  It uses LZ4(HC) internally.  (Analogue of zlib API)

So by nature, LZ4(HC) is lacking many "standard / zlib equivalent" APIs and functionalities.

**Comment by bharatnc** [comment]
@t-mat thank you! I think this will work, let me try it out.

**Comment by bharatnc** [comment]
Thank you @t-mat, here's my update: I was able to use this api to get the compression to work. But when I attempt to do `lz4 -d myfile.lz4`, I get the following error:
```
Error 44 : Unrecognized header : file cannot be decoded 
```
I wonder if this error / behavior is expected and that the file needs to be decoded using similar means as shown in the code example ?

**Comment by t-mat** [comment]
> I wonder if this error / behavior is expected and that the file needs to be decoded using similar means as shown in the code example ?

`frameCompress` example outputs LZ4 Frame Format which is compatible with `lz4` CLI.  You can check it with the following commands.

```sh
cd
git clone https://github.com/lz4/lz4.git lz4-issue-1009
cd lz4-issue-1009
make
cd examples
make

# frameCompress example generates README.md.lz4
./frameCompress README.md
ls -l README.md*

# Use lz4 CLI to decompress README.md.lz4 which is compressed by frameCompress
../programs/lz4 -d README.md.lz4 README.md.lz4.decompressed-lz4cli
cmp README.md README.md.lz4.decompressed-lz4cli
```

> ```
> Error 44 : Unrecognized header : file cannot be decoded
> ```

Usually, this error means wrong magic number.  Please make sure the first 4 bytes of your `.lz4` file is [`0x04, 0x22, 0x4d, 0x18`](https://github.com/lz4/lz4/blob/dev/doc/lz4_Frame_format.md#general-structure-of-lz4-frame-format).

```
od -t x1 -A x -N 4 YOUR-FILE.lz4
```

```
000000 04 22 4d 18
```
